x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -368,80 +469,119 @@ export class ExecutionHandler<
-      notifyUsage: true,
-    });
+  private generateExecutables(
+    alerts: Record<string, Alert<State, Context, ActionGroupIds | RecoveryActionGroupId>>
+  ) {
+    const executables = [];
+
+    for (const action of this.rule.actions) {
+      if (this.isSummaryAction(action)) {
+        if (!this.isSummaryActionThrottled(action)) {
+          executables.push({ action });
+        }
+        continue;
+      }
+      for (const [alertId, alert] of Object.entries(alerts)) {
+        const actionGroup = this.getActionGroup(alert);
+
+        if (!this.ruleTypeActionGroups!.has(actionGroup)) {
+          this.logger.error(
+            `Invalid action group "${actionGroup}" for rule "${this.ruleType.id}".`
+          );
+          continue;
+        }
+        if (action.group === actionGroup && !this.isAlertMuted(alertId)) {
+          if (
+            this.isRecoveredAlert(action.group) ||
+            this.isExecutableActiveAlert({ alert, action })
+          ) {
+            executables.push({ action, alert });
+          }
+        }
+      }
+    }
+
+    return executables;
   }
 
-  private isExecutableActiveAlert({
-    alert,
+  private async getSummarizedAlerts({
     action,
+    ruleId,
+    spaceId,
   }: {
-    alert: Alert<AlertInstanceState, AlertInstanceContext, ActionGroupIds | RecoveryActionGroupId>;
     action: RuleAction;
+    ruleId: string;
+    spaceId: string;
   }) {
-    const alertId = alert.getId();
-    const { rule, ruleLabel, logger } = this;
-    const notifyWhen = action.frequency?.notifyWhen || rule.notifyWhen;
-
-    if (notifyWhen === 'onActionGroupChange' && !alert.scheduledActionGroupHasChanged()) {
-      if (
-        !this.skippedAlerts[alertId] ||
-        (this.skippedAlerts[alertId] &&
-          this.skippedAlerts[alertId].reason !== Reasons.ACTION_GROUP_NOT_CHANGED)
-      ) {
-        logger.debug(
-          `skipping scheduling of actions for '${alertId}' in rule ${ruleLabel}: alert is active but action group has not changed`
-        );
-      }
-      this.skippedAlerts[alertId] = { reason: Reasons.ACTION_GROUP_NOT_CHANGED };
-      return false;
+    let options;
+
+    if (this.isActionOnThrottleInterval(action)) {
+      const throttleMills = parseDuration(action.frequency!.throttle!);
+      const start = new Date(Date.now() - throttleMills);
+
+      options = {
+        start,
+        end: new Date(),
+        ruleId,
+        spaceId,
+      };
+    } else {
+      options = {
+        executionUuid: this.executionId,
+        ruleId,
+        spaceId,
+      };
     }
 
-    if (notifyWhen === 'onThrottleInterval') {
-      const throttled = action.frequency?.throttle
-        ? alert.isThrottled({
-            throttle: action.frequency.throttle ?? null,
-            actionHash: this.generateActionHash(action),
-          })
-        : alert.isThrottled({ throttle: rule.throttle ?? null });
+    const alerts = await this.ruleType.getSummarizedAlerts!(options);
```

Comment:
Not all rule types have `getSummarizedAlerts` defined, so we should check that this function exists for the rule type. The check should probably happen earlier in the execution handler logic.

Reply:
IMHO, the summary option should never be shown on the UI and never be created in the BE if the rule type doesn't have `getSummarizedAlerts` method. That's why i didn't add another check here.

Reply:
That's fair. I don't think we currently have a check in the rule create/update now to verify that a rule type has the `getSummarizedAlerts` function though. Is there an issue to add that?

Reply:
No there isn't. i was planning to add this to the [UI issue](https://github.com/elastic/kibana/issues/143369) since will it show/hide the summary options and use the endpoints.

Reply:
I think we should still add a safety check and log errors gracefully here if for some reason we end up in such a scenario. I tested this locally when `getSummarizedAlerts` is missing and an error gets thrown, subsequent actions in the executables array get skipped.. I'm not sure what else in the post execution logic gets skipped..